tests/spread/plugins/craft-parts/build-and-run-hello/matter-hello/snap/snapcraft.yaml

```diff
@@ -0,0 +1,34 @@
+name: matter-hello
+version: "1.0"
+summary: simple matter application
+description: build an example lighting application using the Matter SDK and say "hello world"
+base: core22
+confinement: strict
+
+apps:
+  matter-hello:
+    command: bin/lighting-app
+
+parts:
+  lighting-hello:
+    plugin: matter
+    matter-sdk-version: "master"
+    matter-zap-version: "v2023.11.13"
+    source: app
+    override-build: |
+      craftctl default
+      cd $CRAFT_PART_BUILD/matter
+
+      # Overwrite the main application file so that it prints "hello world"
+      cp -v $CRAFT_PART_BUILD/main.cpp examples/lighting-app/linux/main.cpp
+
+      # Build the lighting app
+      scripts/examples/gn_build_example.sh examples/lighting-app/linux/ ./out/build
```

Comment:
Building the lighting app from this `gn_build_example.sh` does not require an extra rerun of the `bootstrap.sh`:
```
# Matter plugin
:: Built Matter SDK                                                                                                                                                                  
:: + cd /root/parts/lighting-hello/build/matter      
...                                                                                                                        
:: + cp -v /root/parts/lighting-hello/build/main.cpp examples/lighting-app/linux/main.cpp                                                                                            
:: '/root/parts/lighting-hello/build/main.cpp' -> 'examples/lighting-app/linux/main.cpp'                                                                                             
:: + scripts/examples/gn_build_example.sh examples/lighting-app/linux/ ./out/build                                                                                                   
::     ...                                                  
::   ACTIVATOR! This sets your shell environment variables.                                                                                                                   
::                                                                                                                                                                               
:: Activating environment (setting environment variables):                                                                                                                   
::                                                                                                                                                                                   
::   Setting environment variables for CIPD package manager...done                                                                                                                   
::   Setting environment variables for Project actions........skipped                                                                                                                
::   Setting environment variables for Python environment.....done                                                                                                                   
::   Setting environment variables for pw packages............skipped                                                                                                                
::   Setting environment variables for Host tools.............done                                                                                                                   
::                                                                                                                                                                                   
:: Checking the environment:                                                                                                                                                 
::                                                                                                                                                                                   
:: 20240116 10:34:09 INF Environment passes all checks!                                                                                                                              
::                                                                                                                                                                                   
:: Environment looks good, you are ready to go!                                                                                                                              
::                                                                                                                                                                                   
:: + env                                                                                                                                                                             
...                                                                                                                                                                
:: + gn gen --check --fail-on-unused-args --root=examples/lighting-app/linux/ ./out/build --args=                                                                                    
:: Done. Made 291 targets from 228 files in 161ms                                                                                                                                    
:: + ninja -C ./out/build                                                                                                                                                            
:: ninja: Entering directory `./out/build'      
```

Reply:
This is similar to how we [build the application in Chip Tool snap](https://github.com/canonical/chip-tool-snap/blob/2ad25f14ce58b2f50d1029cc801ef342270e12ed/snap/snapcraft.yaml#L86-L87). It is okay for the tests. However, we still need to identify what additional steps are necessary to build a standalone app that uses the SDK.

Reply:
The `gn_build_example.sh` [calls `scripts/activate.sh`](https://github.com/project-chip/connectedhomeip/blob/master/scripts/examples/gn_build_example.sh#L23), which is a necessary step. The build process would succeed with it, as shown below:
```
set +u && source scripts/activate.sh && set -u

# Build the lighting app
cd examples/lighting-app/linux
gn gen out/build
ninja -C out/build
```
I attempted to call `scripts/activate.sh` in the Matter plugin, but it had no effect. It seems that it has to be called during the app building process.

Reply:
[`scripts/activate.sh`](https://github.com/project-chip/connectedhomeip/blob/09f014cebf114980e6174277d8d45f301f77679f/scripts/activate.sh) is a symbolic link to `scripts/setup/bootstrap.sh`.

So, the reason this script works (compared with using gn and ninja) is that it repeats the bootstrapping. We still need to identify why the bootstrap needs to be repeated, even though it skips most of the steps. Could it be related to the loss of environment variables set in the plugin part?

Logs covering the end of first bootstrap and the second bootstrap:
```
...
:: ++ unset -f _chip_bootstrap_banner                                                                
:: ++ '[' -n '' ']'                                                                                  
:: + set -u                                                                                          
:: + echo 'Built Matter SDK'                                                                         
:: Built Matter SDK                                                                                  
:: + cd /root/parts/lighting-hello/build/matter                                                      
:: + cp -v /root/parts/lighting-hello/build/main.cpp examples/lighting-app/linux/main.cpp            
:: '/root/parts/lighting-hello/build/main.cpp' -> 'examples/lighting-app/linux/main.cpp'             
:: + scripts/examples/gn_build_example.sh examples/lighting-app/linux/ ./out/build                   
::                                                                                            
::   WELCOME TO...                                                                                   
::                                                                                               
::          █                                                                             
::          █                                                                             
::      ▄   █   ▄                                █     █                                  
::      ▀▀█████▀▀      ▄▀▀▀▄ ▄▀▀▀▄    ▄▀▀▀▀▄█  ▀▀█▀▀▀▀▀█▀▀   ▄▀▀▀▀▄    ▄▀▀                
::    ▀█▄       ▄█▀   █     █     █  █      █    █     █    █▄▄▄▄▄▄█  █                   
::      ▀█▄   ▄█▀     █     █     █  █      █    █     █    █         █                   
::   ▄██▀▀█   █▀▀██▄  █     █     █   ▀▄▄▄▄▀█    ▀▄▄   ▀▄▄   ▀▄▄▄▄▀   █                   
::  ▀▀    █   █    ▀▀                                                                     
::                                                                                                   
::   ACTIVATOR! This sets your shell environment variables.                                   
::                                                                                               
:: Activating environment (setting environment variables):                                   
::                                                                                                   
::   Setting environment variables for CIPD package manager...done                                   
::   Setting environment variables for Project actions........skipped                                
::   Setting environment variables for Python environment.....done                                   
::   Setting environment variables for pw packages............skipped                                
::   Setting environment variables for Host tools.............done                                   
::                                                                                                   
:: Checking the environment:                                                                 
::                                                                                                   
:: 20240117 11:08:38 INF Environment passes all checks!                                              
::                                                                                                   
:: Environment looks good, you are ready to go!                                              
::                                                                                                   
:: + env                                                                                             
:: PYLXD_WARNINGS=none                                                                               
:: CRAFT_PART_SRC_WORK=/root/parts/lighting-hello/src                                                
:: SNAPCRAFT_TARGET_ARCH=amd64                                                                       
:: SNAP_REVISION=x1                                                                                  
:: SNAP_REAL_HOME=/root                                                                              
:: SNAP_USER_COMMON=/root/snap/snapcraft/common                                                      
:: PW_PROJECT_ROOT=/root/parts/lighting-hello/build/matter                                           
:: SNAP_INSTANCE_KEY=                                                                                
:: SNAPCRAFT_PROJECT_GRADE=                                                                          
:: PW_ZAP_CIPD_INSTALL_DIR=/root/parts/lighting-hello/build/matter/.environment/cipd/packages/zap    
:: SNAP_EUID=0                                                                                       
:: PWD=/root/parts/lighting-hello/build/matter                                                       
:: SNAPCRAFT_PROJECT_NAME=matter-hello                                                               
:: CRAFT_PART_SRC=/root/parts/lighting-hello/src                                                     
:: TZ=Europe/Berlin                                                                                  
:: CRAFT_ARCH_BUILD_ON=amd64                                                                         
:: CIPD_CACHE_DIR=/root/.cipd-cache-dir                                                              
:: SNAP_CONTEXT=JDcECUvOasWnHF82QWnLIxt_cvnTeqVuIsisYyNo5zcHXQZKVvgv                                 
:: container=lxc                                                                                     
:: PW_ROOT=/root/parts/lighting-hello/build/matter/third_party/pigweed/repo                          
:: HOME=/root                                                                                        
:: SNAPCRAFT_PROJECT_DIR=/root/project                                                               
:: LANG=C.UTF-8                                                                                      
:: CRAFT_PROJECT_DIR=/root/project                                                                   
:: CRAFT_PART_BUILD=/root/parts/lighting-hello/build                                                 
:: VIRTUAL_ENV=/root/parts/lighting-hello/build/matter/.environment/pigweed-venv                     
:: SNAP_ARCH=amd64                                                                                   
:: SNAP_INSTANCE_NAME=snapcraft                                                                      
:: SNAP_USER_DATA=/root/snap/snapcraft/x1                                                            
:: SNAPCRAFT_MANAGED_MODE=1                                                                          
:: SNAPCRAFT_PART_BUILD=/root/parts/lighting-hello/build                                             
:: SNAP_REEXEC=                                                                                      
:: SNAP_UID=0                                                                                        
:: PW_PYTHON_CIPD_INSTALL_DIR=/root/parts/lighting-hello/build/matter/.environment/cipd/packages/python                                                                                                    
:: SNAPCRAFT_PART_INSTALL=/root/parts/lighting-hello/install                                         
:: _PW_ROSETTA=0                                                                                     
:: CRAFT_PROJECT_NAME=matter-hello                                                                   
:: CRAFT_PART_BUILD_WORK=/root/parts/lighting-hello/build                                            
:: SNAPCRAFT_PARALLEL_BUILD_COUNT=16                                                                 
:: SNAPCRAFT_PART_BUILD_WORK=/root/parts/lighting-hello/build                                        
:: CRAFT_ARCH_TRIPLET_BUILD_FOR=x86_64-linux-gnu                                                     
:: TERM=xterm-256color                                                                               
:: CRAFT_ARCH_TRIPLET=x86_64-linux-gnu                                                               
:: USER=root                                                                                         
:: SNAPCRAFT_ARCH_TRIPLET=x86_64-linux-gnu                                                           
:: CRAFT_STEP_NAME=BUILD                                                                             
:: SNAP=/snap/snapcraft/x1                                                                           
:: SNAPCRAFT_PROJECT_VERSION=1.0                                                                     
:: SNAP_COMMON=/var/snap/snapcraft/common                                                            
:: SNAP_VERSION=6.0.post983+git1e5909bd                                                              
:: SHLVL=2                                                                                           
:: SNAP_LIBRARY_PATH=/var/lib/snapd/lib/gl:/var/lib/snapd/lib/gl32:/var/lib/snapd/void               
:: SNAP_COOKIE=JDcECUvOasWnHF82QWnLIxt_cvnTeqVuIsisYyNo5zcHXQZKVvgv                                  
:: PW_PACKAGE_ROOT=/root/parts/lighting-hello/build/matter/.environment/packages                     
:: CRAFT_PARALLEL_BUILD_COUNT=16                                                                     
:: SNAPCRAFT_PART_SRC=/root/parts/lighting-hello/src                                                 
:: PARTS_CALL_FIFO=/tmp/tmpecewbrjw/function_call                                                    
:: _PW_ENVIRONMENT_CONFIG_FILE=/root/parts/lighting-hello/build/matter/scripts/setup/environment.json
:: SNAP_DATA=/var/snap/snapcraft/x1                                                                  
:: CRAFT_PRIME=/root/prime                                                                           
:: LD_LIBRARY_PATH=/snap/snapcraft/x1/none                                                           
:: _PW_ACTUAL_ENVIRONMENT_ROOT=/root/parts/lighting-hello/build/matter/.environment                  
:: PW_BRANDING_BANNER=/root/parts/lighting-hello/build/matter/scripts/setup/banner.txt               
:: SNAPCRAFT_STAGE=/root/stage                                                                       
:: CRAFT_ARCH_TRIPLET_BUILD_ON=x86_64-linux-gnu                                                      
:: PW_ARM_CIPD_INSTALL_DIR=/root/parts/lighting-hello/build/matter/.environment/cipd/packages/arm    
:: CRAFT_TARGET_ARCH=amd64                                                                           
:: PARTS_FEEDBACK_FIFO=/tmp/tmpecewbrjw/call_feedback                                                
:: CRAFT_STAGE=/root/stage                                                                           
:: SNAPCRAFT_PART_SRC_WORK=/root/parts/lighting-hello/src                                            
:: CRAFT_PART_NAME=lighting-hello                                                                    
:: SNAP_NAME=snapcraft                                                                               
:: PATH=/root/parts/lighting-hello/build/matter/third_party/pigweed/repo/out/host/host_tools:/root/parts/lighting-hello/build/matter/.environment/pigweed-venv/bin:/root/parts/lighting-hello/build/matter/.environment/cipd/packages/arm/bin:/root/parts/lighting-hello/build/matter/.environment/cipd/packages/arm:/root/parts/lighting-hello/build/matter/.environment/cipd/packages/zap:/root/parts/lighting-hello/build/matter/.environment/cipd/packages/pigweed/bin:/root/parts/lighting-hello/build/matter/.environment/cipd/packages/pigweed:/root/parts/lighting-hello/build/matter/.environment/cipd:/snap/snapcraft/x1/libexec/snapcraft:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin              
:: CRAFT_PART_INSTALL=/root/parts/lighting-hello/install                                             
:: SNAPCRAFT_PRIME=/root/prime                                                                       
:: CRAFT_ARCH_BUILD_FOR=amd64                                                                        
:: PW_PIGWEED_CIPD_INSTALL_DIR=/root/parts/lighting-hello/build/matter/.environment/cipd/packages/pigweed                                                                                                  
:: OLDPWD=/root/parts/lighting-hello/build                                                           
:: _=/usr/bin/env                                                                                    
:: + gn gen --check --fail-on-unused-args --root=examples/lighting-app/linux/ ./out/build --args=    
:: Done. Made 292 targets from 228 files in 105ms                                                    
:: + ninja -C ./out/build                                                                            
:: ninja: Entering directory `./out/build'                                                           
:: [1/597] stamp obj/third_party/connectedhomeip/src/app/icd/observer.stamp                          
:: [2/597] stamp obj/third_party/connectedhomeip/src/app/revision_info.stamp     
...
```

Reply:
After Comparing the environment variables in the Matter plugin with the lighting-hello part, it appears that all pigweed related environment variables set by bootstrap are missing. There could be reasons such as the plugin being a child or a separate process of the part. 

Potential solutions could include exporting the plugin's environment to an env file and sourcing it from the part. Alternatively, need to find a way to make the plugin and the part share the same environment.

Reply:
ok. Given the complexity remained on the consumer side (layout mapping, and now env file sourcing), we may end up requiring an extension too. 

Given the environment differences, the fact that the consumer always has to use override-build to build the app (the default is never enough), and the [plugin caching issue](https://github.com/snapcore/snapcraft/pull/4491#pullrequestreview-1782100684), maybe the plugin design should change to make it a "matter-sdk" plugin, used exclusively in a standalone part. What do you think?

Reply:
For now, placing the Matter sdk plugin as a standalone `part` and passing the environment to the next `part` by sourcing an environment file is the easiest approach.

Reply:
There are two ways for the matter-lighting `part` to source environment variables:
- First, store environment variables in an env file within the matter plugin. Then, the consumer `part` sources the env file
- On the consumer side only: Use this script `scripts/examples/gn_build_example.sh` to build the application, which includes a bootstrap to set up environment variables.

I prefer the second approach. In both approaches, the consumer side needs to perform an additional environment source. However, the second approach only requires a change on the consumer side, whereas the first one require changes on both the plugin and consumer sides.

Reply:
There are two issues with the second approach:
- The plugin isn't meant to just build examples from the SDK. The consumer should be able to build a standalone application that uses the SDK.
- The environment variables needed to build aren't just the ones set by the bootstrap script, but also those set additionally by the plugin.

Reply:
Then the first approach is better: store environment variables in an env file within the matter plugin. Then, the consumer part sources the env file